source/aws_s3/s3.go

```diff
@@ -21,37 +21,64 @@ func init() {
 
 type s3Driver struct {
 	s3client   s3iface.S3API
-	bucket     string
-	prefix     string
+	config     *Config
 	migrations *source.Migrations
 }
 
+type Config struct {
+	Bucket string
+	Prefix string
+}
+
 func (s *s3Driver) Open(folder string) (source.Driver, error) {
-	u, err := url.Parse(folder)
+	sess, err := session.NewSession()
 	if err != nil {
 		return nil, err
 	}
-	sess, err := session.NewSession()
+
+	config, err := parseURI(folder)
 	if err != nil {
 		return nil, err
 	}
-	driver := s3Driver{
-		bucket:     u.Host,
-		prefix:     strings.Trim(u.Path, "/") + "/",
-		s3client:   s3.New(sess),
+
+	return WithInstance(s3.New(sess), config)
+}
+
+func WithInstance(s3client s3iface.S3API, config *Config) (source.Driver, error) {
+	driver := &s3Driver{
+		config:     config,
+		s3client:   s3client,
 		migrations: source.NewMigrations(),
 	}
-	err = driver.loadMigrations()
+
+	if err := driver.loadMigrations(); err != nil {
+		return nil, err
+	}
+
+	return driver, nil
+}
+
+func parseURI(uri string) (*Config, error) {
+	u, err := url.Parse(uri)
 	if err != nil {
 		return nil, err
 	}
-	return &driver, nil
+
+	prefix := strings.Trim(u.Path, "/")
```

Comment:
I don't think trimming is necessary.
e.g. specify your bucket as `s3://my-bucket` instead of `s3://my-bucket/`
We can update the docs accordingly

Reply:
Unfortunately, trimming is necessary. If you specify a prefix (ex `s3://migrations-bucket/production`),  `u.Path` is `/production` and we need to strip off the first slash and add a trailing slash to get `production/`.

Additionally, if you added a trailing slash (ex `s3://migrations-bucket/production/`), `u.Path` ends up being `/production/` which we need to convert to `production/`.

The string trimming was in the original code, I dont think it makes much sense to remove it and risk breaking backwards compatibility.